clj-refactor.el

```diff
@@ -1983,37 +1983,68 @@ the alias in the project."
+  (assert (not (completable-for-cljr-slash? nil)))
+  (assert (completable-for-cljr-slash? "a"))
+  (assert (completable-for-cljr-slash? "a-"))
+  (assert (completable-for-cljr-slash? "a2"))
+  (assert (completable-for-cljr-slash? "a-2"))
+  (assert (completable-for-cljr-slash? "a2-"))
+  (assert (completable-for-cljr-slash? "a2.a"))
+  (assert (completable-for-cljr-slash? "a2.a-"))
+  (assert (completable-for-cljr-slash? "a2.a2"))
+  (assert (completable-for-cljr-slash? "a2.a-2"))
+  (assert (completable-for-cljr-slash? "a2.a2-"))
+  (assert (not (completable-for-cljr-slash? "a2.a.")))
+  (assert (not (completable-for-cljr-slash? "a2.2")))
+  (assert (not (completable-for-cljr-slash? "a2.2a")))
+  (assert (not (completable-for-cljr-slash? "a2.-")))
+  (assert (not (completable-for-cljr-slash? "a2.-a")))
+  (assert (not (completable-for-cljr-slash? "-")))
+  (assert (not (completable-for-cljr-slash? ".")))
+  (assert (not (completable-for-cljr-slash? "2")))
+  (assert (not (completable-for-cljr-slash? "+")))
+  (assert (not (completable-for-cljr-slash? "a/")))
+  (assert (not (completable-for-cljr-slash? "a/a")))
+  (assert (not (completable-for-cljr-slash? "a/a/"))))
+
 ;;;###autoload
 (defun cljr-slash ()
   "Inserts / as normal, but also checks for common namespace shorthands to require.
 If `cljr-magic-requires' is non-nil, executing this command after one of the aliases
 listed in `cljr-magic-require-namespaces', or any alias used elsewhere in the project,
 will add the corresponding require statement to the ns form."
   (interactive)
-  (insert "/")
-  (when-let (aliases (and cljr-magic-requires
-                          (not (cljr--in-map-destructuring?))
-                          (not (cljr--in-ns-above-point-p))
-                          (not (cljr--in-reader-literal-p))
-                          (not (cider-in-comment-p))
-                          (not (cider-in-string-p))
-                          (not (cljr--in-keyword-sans-alias-p))
-                          (not (cljr--in-number-p))
-                          (clojure-find-ns)
-                          (cljr--magic-requires-lookup-alias)))
-    (let ((short (cl-first aliases)))
-      (when-let (long (cljr--prompt-user-for "Require " (cl-second aliases)))
-        (when (and (not (cljr--in-namespace-declaration-p (concat ":as " short "\b")))
-                   (or (not (eq :prompt cljr-magic-requires))
-                       (not (> (length (cl-second aliases)) 1)) ; already prompted
-                       (yes-or-no-p (format "Add %s :as %s to requires?" long short))))
-          (save-excursion
-            (cljr--insert-in-ns ":require")
-            (let ((libspec (format "[%s :as %s]" long short)))
-              (insert libspec)
-              (ignore-errors (cljr--maybe-eval-ns-form))
-              (cljr--indent-defun)
-              (cljr--post-command-message "Required %s" libspec))))))))
+  (let ((original-input (cider-symbol-at-point)))
+    (insert "/")
```

Comment:
if we insert the `/` before anything else, the `cider-symbol-at-point` (`thing-at-point` now) query will be affected by that mutation. So it will be harder to get an accurate answer

Reply:
Yeah, but we've had something that's worked pretty well for years, that we can use instead of `thing-at-point`:

```el
(thread-last (buffer-substring-no-properties
                             (cljr--point-after 'paredit-backward)
                             (1- (point)))
                 (string-remove-prefix "::")
                 (string-remove-prefix "@"))
```

Though looking at this, after you mentioned metadata, I see that it might not handle that properly.

Reply:
`thing-at-point` is working nicely, thanks!

The code you quote is also useful for the next step, I refined it in one of the commits (`Extract and complete cljr--namespace-of-thing-at-point`)